control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs

```diff
@@ -0,0 +1,146 @@
+use crate::controller::{
+    reconciler::{PollContext, TaskPoller},
+    task_poller::{PollEvent, PollResult, PollTimer, PollerState},
+};
+use agents::errors::SvcError;
+use std::time::Duration;
+use tokio::time::sleep;
+
+use crate::controller::resources::operations::ResourcePublishing;
+use common_lib::types::v0::{
+    store::node::NodeSpec,
+    transport::{NodeId, RepublishVolume, VolumeId, VolumeShareProtocol},
+};
+
+/// Node drain reconciler.
+#[derive(Debug)]
+pub(super) struct NodeNexusReconciler {
+    counter: PollTimer,
+}
+impl NodeNexusReconciler {
+    /// Return a new `Self`.
+    pub(super) fn new() -> Self {
+        Self {
+            counter: PollTimer::from(1), // sets the reconciler polling rate
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl TaskPoller for NodeNexusReconciler {
+    async fn poll(&mut self, context: &PollContext) -> PollResult {
+        let mut results = vec![];
+        let nodes = context.specs().get_nodes();
+
+        for node in nodes {
+            results.push(check_and_drain_node(context, &node).await);
+        }
+        Self::squash_results(results)
+    }
+
+    async fn poll_timer(&mut self, _context: &PollContext) -> bool {
+        self.counter.poll()
+    }
+
+    async fn poll_event(&mut self, context: &PollContext) -> bool {
+        matches!(context.event(), PollEvent::TimedRun)
+    }
+}
+
+/// Republish the specified volume on any node other than its current node
+async fn republish_volume(context: &PollContext, vol_uuid: &VolumeId) -> Result<(), SvcError> {
+    let request = RepublishVolume::new(vol_uuid.clone(), None, VolumeShareProtocol::Nvmf, false);
+    tracing::info!(
+        volume.uuid = vol_uuid.as_str(),
+        "Attempting to republish volume"
+    );
+
+    let mut volume = context.registry().specs().volume(&request.uuid).await?;
+    volume.republish(context.registry(), &request).await?;
+
+    Ok(())
+}
+
+/// Returns true if the specified node is in the draining state
+fn is_node_draining(context: &PollContext, nodeid: &NodeId) -> Result<bool, SvcError> {
+    let nodespec = match context.specs().get_node(nodeid) {
+        Ok(ns) => ns,
+        Err(e) => {
+            return Err(e);
+        }
+    };
+    Ok(nodespec.is_draining())
+}
+
+/// Drain the specified node if in draining state
+async fn check_and_drain_node(context: &PollContext, node_spec: &NodeSpec) -> PollResult {
+    if node_spec.is_draining() {
+        let node_id = node_spec.id().clone();
+        tracing::trace!(node.id = node_spec.id().as_str(), "Draining node");
+        // iterate through all the volumes, find those with a nexus hosted on the node
+        let volumes = context.registry().get_volumes().await;
+
+        for volume in volumes.iter() {
+            if volume.spec().target.is_some()
+                && volume.spec().target.unwrap().node().clone() == node_id
```

Comment:
```suggestion
            let target = match volume.spec().target() { Some(t) => t, None => continue };
            if target.node() == node_id
```

Reply:
That part of the code was rewritten in the form: if let Some(target) = ....